code/modules/mob/living/simple_animal/hostile/necro.dm

```diff
@@ -98,11 +124,295 @@
+	playsound(get_turf(D), 'sound/effects/grillehit.ogg', 50, 1)
+	D.shake(1, 8)
+	stop_automated_movement = 1
+	spawn(100) //Roughly 10 seconds for people on the other side of the door to run the heck away
+		if(can_open_door(D))//Let's see if nobody quickly bolted it
+			if(istype (src, /mob/living/simple_animal/hostile/necro/zombie/crimson)) //Guaranteed
+				D.visible_message("<span class='warning'>\The [D] breaks open under the pressure</span>")
+				if(istype(D, /obj/machinery/door/airlock/))
+					var/obj/machinery/door/airlock/A = D
+					A.locked = 0
+					A.welded = 0
+					A.jammed = 0
+				D.open(1)
+			else
+				if(prob(15))
+					D.visible_message("<span class='warning'>\The [D] creaks open under force, steadily</span>")
+					D.open(1)
+		busy = 0
+		stop_automated_movement = 0
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/check_edibility(var/mob/living/carbon/human/target)
+	if(isjusthuman(target)) //Humans are always edible
+		return 1
+	if(target.health > -400) //So they're not caught eating the same dumb bird all day
+		return 1
+
+	return 0
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/eat(var/mob/living/carbon/human/target)
+	//Deal a random amount of brute damage to the corpse in question, heal the zombie by the damage dealt halved
+	visible_message("<span class='notice'>\the [src] starts to take a bite out of \the [target].</span>")
+	spawn(50)
+		playsound(get_turf(src), 'sound/weapons/bite.ogg', 50, 1)
+		var/damage = rand(melee_damage_lower, melee_damage_upper)
+		target.adjustBruteLoss(damage)
+		health += (damage/2)
+		if(maxHealth < health_cap)
+			maxHealth += 5 //A well fed zombie is a scary zombie
+		times_eaten += 1
+		busy = 0
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/check_evolve()
+	if(!can_evolve) //How did you get here if not?
+		return
+
+	/*
+					Turned (Just reanimated, can be turned back)
+										V
+				Rotting (If eaten once, or died once, can't be turned back)
+				/													\
+			Putrid													Crimson
+	Eaten too much, died too little								Eaten too little, died too much
+	*/
+	if(istype(src, /mob/living/simple_animal/hostile/necro/zombie/turned))
+		if(times_revived > 0 || times_eaten > 0)
+			evolve(/mob/living/simple_animal/hostile/necro/zombie/rotting)
+
+	else if (istype(src, /mob/living/simple_animal/hostile/necro/zombie/rotting))
+		if(times_eaten > (1+times_revived)*2) //Have to have eaten at least twice and more than double that of times died
+			evolve(/mob/living/simple_animal/hostile/necro/zombie/putrid)
+		else if(times_revived > times_eaten+1) //Died at least twice
+			evolve(/mob/living/simple_animal/hostile/necro/zombie/crimson)
+
+	busy = 0
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/evolve(var/mob/living/simple_animal/evolve_to)
+	if(istype(evolve_to, /mob/living/simple_animal/hostile/necro))
+		var/mob/living/evolution = new evolve_to(src.loc,,)
+		evolution.name = name //We want to keep the name
```

Comment:
Now, for some weird reason, it's not actually keeping the name and I'm not entirely sure why